### PR TITLE
Selectively disable annotations

### PR DESCRIPTION
**Pull request by JustinHoldstock**
Annotations can now be selectively disabled through preview options. 
Viewer options override global `showAnnotations` value, for that viewer.
Tests to follow.
```javascript
// Turn on/off annotations for all viewers
preview.show(..., {
    showAnnotations: Boolean
});
```
Combined with the following:
```javascript
// [LEGACY] BOOLEAN - turn on/off annotations for a viewer
preview.show(..., {
    viewers: {
        VIEWER_NAME: {
            annotations: Boolean
        }
    }
});

// [NEW] OBJECT - turn on/off annotations for a viewer.
preview.show(..., {
    viewers: {
        VIEWER_NAME: {
            annotations: {
                enabled: Boolean, // Enables/disables if set. Respects "showAnnotations" if empty
                disabledTypes: String[] // List of annotation types to disable
            }
        }
    }
});
```

And an example. Enable all annotations, turn off for Image Viewers, and
disable `point` annotations on Document viewer:

```javascript
preview.show(fileId, token, {
    showAnnotations: true,
    viewers: {
        Image: {
            annotations: {
                enabled: false
            }
        },
        Document: {
            annotations: {
                disabledTypes: ['point']
            }
        }
    }
});
```

### CODE REVIEW

File: src/lib/annotations/BoxAnnotations.js [hunk 1]
```diff
@@ -37,16 +37,48 @@ class BoxAnnotations {
...
+     * @param {Array} [viewerConfig] - Annotation configuration for a specific viewer
+     * @return {Object|null} A copy of the annotator to use, if available
      */
-    determineAnnotator(viewer, disabledAnnotators = []) {
-        return this.annotators.find(
-            (annotator) => !disabledAnnotators.includes(annotator.NAME) && annotator.VIEWER.includes(viewer)
-        );
+    determineAnnotator(viewerName, disabledAnnotators = [], viewerConfig = {}) {
+        const originalAnnotator = this.getAnnotatorsForViewer(viewerName, disabledAnnotators);
+
```
Comment (pramodsum) [inline src/lib/annotations/BoxAnnotations.js]:
```
const annotator = this.getAnnotatorsForViewer(viewerName, disabledAnnotators);
const modifiedAnnotator = null;
if (!annotator) {
    return modifiedAnnotator;
}

modifiedAnnotator = ....
if (viewerConfig.enabled) {
    if (Array.isArray(....)) {
        ...
    }
}
return modifiedAnnotator;
```

File: src/lib/annotations/BoxAnnotations.js [hunk 2]
```diff
@@ -37,16 +37,48 @@ class BoxAnnotations {
...
      * @param {Array} [disabledAnnotators] - List of disabled annotators
-     * @return {Object} The annotator to use
+     * @param {Array} [viewerConfig] - Annotation configuration for a specific viewer
+     * @return {Object|null} A copy of the annotator to use, if available
      */
-    determineAnnotator(viewer, disabledAnnotators = []) {
-        return this.annotators.find(
-            (annotator) => !disabledAnnotators.includes(annotator.NAME) && annotator.VIEWER.includes(viewer)
-        );
+    determineAnnotator(viewerName, disabledAnnotators = [], viewerConfig = {}) {
```
Comment (pramodsum) [inline src/lib/annotations/BoxAnnotations.js]: Can we do this instead
```determineAnnotator(viewerName, viewerConfig = {}, disabledAnnotators = []) {```

File: src/lib/viewers/BaseViewer.js [hunk 3]
```diff
@@ -623,27 +623,51 @@ class BaseViewer extends EventEmitter {
...
+        }
+
         /* global BoxAnnotations */
         const boxAnnotations = new BoxAnnotations();
+        // #TODO(@spramod|@jholdstock): remove this after we have annotation instancing
+        const viewerName = this.options.viewer.NAME;
+        const annotationsConfig = this.getViewerAnnotationsConfig();
 
-        this.annotatorConf = boxAnnotations.determineAnnotator(this.options.viewer.NAME);
+        this.annotatorConf = boxAnnotations.determineAnnotator(viewerName, undefined, annotationsConfig);
```
Comment (pramodsum) [inline src/lib/viewers/BaseViewer.js]: See comment above but if we modify the function declaration, then we can call it like this
```boxAnnotations.determineAnnotator(viewerName, annotationsConfig);```

File: src/lib/annotations/BoxAnnotations.js (hunk 2 above)
Comment (JustinHoldstock) [inline src/lib/annotations/BoxAnnotations.js]: +1, and done

File: src/lib/viewers/BaseViewer.js (hunk 3 above)
Comment (JustinHoldstock) [inline src/lib/viewers/BaseViewer.js]: DONE! :D

File: src/lib/annotations/BoxAnnotations.js (hunk 1 above)
Comment (JustinHoldstock) [inline src/lib/annotations/BoxAnnotations.js]: Much better, ty. Done

File: src/lib/annotations/BoxAnnotations.js [hunk 4]
```diff
@@ -55,31 +55,32 @@ class BoxAnnotations {
      * Chooses a annotator based on viewer.
      *
      * @param {Object} viewerName - Current preview viewer name
-     * @param {Array} [disabledAnnotators] - List of disabled annotators
      * @param {Array} [viewerConfig] - Annotation configuration for a specific viewer
+     * @param {Array} [disabledAnnotators] - List of disabled annotators
```
Comment (pramodsum) [inline src/lib/annotations/BoxAnnotations.js]: reorder these

File: src/lib/annotations/BoxAnnotations.js [hunk 5]
```diff
@@ -55,31 +55,32 @@ class BoxAnnotations {
...
+            return modifiedAnnotator;
+        }
 
-            // Filter out disabled annotation types
-            if (Array.isArray(viewerConfig.disabledTypes)) {
-                annotator.TYPE = annotator.TYPE.filter((type) => {
-                    return !viewerConfig.disabledTypes.includes(type);
-                });
-            }
+        modifiedAnnotator = Object.assign({}, annotator);
```
Comment (pramodsum) [inline src/lib/annotations/BoxAnnotations.js]: Can this be moved below and then you can just do `return modifiedAnnotator;` below?

File: src/lib/viewers/BaseViewer.js [hunk 6]
```diff
@@ -628,7 +628,8 @@ class BaseViewer extends EventEmitter {
      * @return {Object} An object containing configuration properties.
      */
     getViewerAnnotationsConfig() {
-        const config = this.getViewerOption('annotations') || {};
+        const option = this.getViewerOption('annotations');
+        const config = option !== null && option !== undefined ? option : {};
```
Comment (pramodsum) [inline src/lib/viewers/BaseViewer.js]: `const config = (option !== null && option !== undefined) ? option : {};`

but also does the previous version not work? 
`const config = this.getViewerOption('annotations') || {};`

File: src/lib/annotations/BoxAnnotations.js (hunk 4 above)
Comment (JustinHoldstock) [inline src/lib/annotations/BoxAnnotations.js]: This already has been reorder to what we've already discussed

File: src/lib/annotations/BoxAnnotations.js (hunk 5 above)
Comment (JustinHoldstock) [inline src/lib/annotations/BoxAnnotations.js]: Done

File: src/lib/viewers/BaseViewer.js (hunk 6 above)
Comment (JustinHoldstock) [inline src/lib/viewers/BaseViewer.js]: Previous doesn't work because the option can return `false`

Comment (JustinHoldstock) [inline src/lib/viewers/BaseViewer.js]: done

File: src/lib/annotations/BoxAnnotations.js (hunk 4 above)
Comment (pramodsum) [inline src/lib/annotations/BoxAnnotations.js]: 👍

**Pull request merged**